File: src/features/library/merge.ts
/* MISSION-089 — Merge data layer. Planning is a read; applying folds the
   duplicate into the survivor and parks an undo image in the trash. */

import { useMutation, useQueryClient } from "@tanstack/react-query";
import { merge_apply, merge_plan, type MergePreview, type MergeResult } from "@/api";
import { queryKeys } from "@/api";

/** Preview what merging the duplicate into the survivor would change. */
export function useMergePlan() {
  return useMutation({
    mutationFn: (input: { survivorId: string; duplicateId: string }): Promise<MergePreview> =>
      merge_plan({ survivor_id: input.survivorId, duplicate_id: input.duplicateId }),
  });
}

/** Apply a merge; the trash entry in the result is the undo handle. */
export function useMergeApply() {
  const client = useQueryClient();
  return useMutation({
    mutationFn: (input: { survivorId: string; duplicateId: string }): Promise<MergeResult> =>
      merge_apply({ survivor_id: input.survivorId, duplicate_id: input.duplicateId }),
    onSuccess: () => {
      void client.invalidateQueries({ queryKey: queryKeys.media.all() });
      void client.invalidateQueries({ queryKey: queryKeys.trash.all() });
    },
  });
}

export type { MergePreview, MergeResult };

File: src-tauri/src/commands/merge.rs
//! Merge commands (MISSION-089). Thin handlers — planning and application
//! live in `application::merge_service`; undo lives in the trash layer.

use sqlx::SqlitePool;
use tauri::command;
use tauri::State;
use tracing::info;

use crate::application::merge_service::{MergePreview, MergeResult, MergeService};
use crate::error::AppError;

/// Preview what merging `duplicate_id` into `survivor_id` would change:
/// field conflicts, moved nodes / review / tracking / collections.
#[command]
pub async fn merge_plan(
    state: State<'_, SqlitePool>,
    survivor_id: String,
    duplicate_id: String,
) -> Result<MergePreview, AppError> {
    info!("merge_plan invoked");
    let service = MergeService::new(state.inner().clone());
    service.plan(&survivor_id, &duplicate_id).await
}

/// Apply a merge: snapshot the duplicate into trash (undoable), fold its
/// data into the survivor and delete it.
#[command]
pub async fn merge_apply(
    state: State<'_, SqlitePool>,
    survivor_id: String,
    duplicate_id: String,
) -> Result<MergeResult, AppError> {
    info!("merge_apply invoked");
    let service = MergeService::new(state.inner().clone());
    service.apply(&survivor_id, &duplicate_id).await
}
